File: cpp/0872_leafSimilar.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
// dfs两棵树，cpp里面两个vector是有重载==运算符的！
class Solution {
public:
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
        vector<int> r1,r2;
        traverse(root1,r1);
        traverse(root2,r2);
        return r1==r2;
    }
    void traverse(TreeNode* root,vector<int> &record)
    {
        if(root==nullptr)
            return;
        if(root->left==nullptr && root->right==nullptr)
            record.push_back(root->val);
        traverse(root->left,record);
        traverse(root->right,record);
    }
};

File: java/0872_leafSimilar.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        List<Integer> r1 = new ArrayList<Integer>();
        List<Integer> r2 = new ArrayList<Integer>();
        traverse(root1,r1);
        traverse(root2,r2);
        if(r1.size()!=r2.size())
            return false;
        for(var i=0;i<r1.size();i++)
            if(r1.get(i)!=r2.get(i))
                return false;
        return  true;
    }
    public void traverse(TreeNode root,List<Integer> record)
    {
        if(root==null)
            return;
        if(root.left==null && root.right==null)
            record.add(root.val);
        traverse(root.left,record);
        traverse(root.right,record);
    }
}
